src/inject/utils/dom.ts

```diff
@@ -145,3 +145,153 @@ export function iterateShadowNodes(root: Node, iterator: (node: Element) => void
         iterateShadowNodes(node.shadowRoot, iterator);
     }
 }
+
+export function isDOMReady() {
+    return document.readyState === 'complete' || document.readyState === 'interactive';
+}
+
+const readyStateListeners = new Set<() => void>();
+
+export function addDOMReadyListener(listener: () => void) {
+    readyStateListeners.add(listener);
+}
+
+export function removeDOMReadyListener(listener: () => void) {
+    readyStateListeners.delete(listener);
+}
+
+if (!isDOMReady()) {
+    const onReadyStateChange = () => {
+        if (isDOMReady()) {
+            document.removeEventListener('readystatechange', onReadyStateChange);
+            readyStateListeners.forEach((listener) => listener());
+            readyStateListeners.clear();
+        }
+    };
+    document.addEventListener('readystatechange', onReadyStateChange);
+}
+
+const HUGE_MUTATIONS_COUNT = 1000;
+
+function isHugeMutation(mutations: MutationRecord[]) {
+    if (mutations.length > HUGE_MUTATIONS_COUNT) {
+        return true;
+    }
+
+    let addedNodesCount = 0;
+    for (let i = 0; i < mutations.length; i++) {
```

Comment:
Length caching

Reply:
Let's trash what? What's the point of length caching?

Reply:
It was a joke :) 
But ![](https://i.imgur.com/22SmDhP.png)
https://jsben.ch/kidOp 
Just that little performance boost

Reply:
The performance boost can't be little :)

Here's what I see in Firefox
<img width="600" alt="Screen Shot 2020-05-19 at 19 38 03" src="https://user-images.githubusercontent.com/9638588/82353659-68fc4300-9a08-11ea-9eee-74dc8da8b7bd.png">
and after a 3rd relaunch in Chrome
<img width="583" alt="Screen Shot 2020-05-19 at 19 37 11" src="https://user-images.githubusercontent.com/9638588/82353748-7f0a0380-9a08-11ea-9632-b9b2fdcf56de.png">
So, it's obvious that `for..of` is 4 times slower in Firefox and 25% slower in Chrome. But all other functions are equal, and I choose what looks clearer (though some people prefer to use `Array.some` or `Array.every` to exit loop earlier, I think `for` loop is more clear).

Reply:
I don't get at all the same result........
Poor laptop

Reply:
... And maybe you've seen this saying of Donald Knuth
> Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%

Reply:
If every time someone says `BUT THIS DROP PERFORMANCE` etc it start to get worried and I didn't though about performance at all if you look at my earliest PR's for darkreader with changes of codebase I used var =. I used a lot of functions what easy could be reduced to 1 etc.